File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.cc
#include <iostream>
#include <cmath>

using namespace std;

void space(int n);
void recur(int n);

int main(void) {
	while (true) {
		int n;

		cin >> n;

		if (cin.eof())
			break;

		recur(n);
		cout << '\n';
	}
	return 0;
}

void space(int n) {
	int repeat = pow(3, n);

	while (repeat--) {
		cout << ' ';
	}
}

void recur(int n) {
	if (n == 0) {
		cout << '-';
		return;
	}

	recur(n - 1);
	space(n - 1);
	recur(n - 1);
}

File: 백준/Silver III/4779. 칸토어 집합/칸토어 집합.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while(sc.hasNext()){
            int n = sc.nextInt();

            StringBuilder sb = new StringBuilder();
            recur(sb, n);
            System.out.println(sb);
        }
    }

    public static void recur(StringBuilder sb, int n) {
        if(n == 0){
            sb.append('-');
            return;
        }

        recur(sb, n - 1);
        space(sb, n - 1);
        recur(sb, n - 1);
    }

    public static void space(StringBuilder sb, int n) {
        int repeat = (int)Math.pow(3, n);

        while(repeat-- > 0) {
            sb.append(' ');
        }
    }
}
